onnxruntime/test/perftest/performance_runner.cc

```diff
@@ -129,32 +134,38 @@ Status PerformanceRunner::RunParallelDuration() {
 }
 
 Status PerformanceRunner::ForkJoinRepeat() {
-  // Adding trivially simple parallelization to the repeated times test will simply perform
-  // m instances of n parallel invocations with a synchronized join after each invocation.
-  // TODO: When the thread pool implementation is done, redo if it has join semantics.
-  auto tpool = GetDefaultThreadPool(Env::Default());
-  std::atomic<int> counter = {0};
+  const auto& run_config = performance_test_config_.run_config;
+
+  // create a threadpool with one thread per concurrent request
+  auto tpool = std::make_unique<DefaultThreadPoolType>(run_config.concurrent_session_runs);
+  std::atomic<int> counter{0}, requests{0};
   std::mutex m;
   std::condition_variable cv;
 
-  for (size_t ite = 0; ite < performance_test_config_.run_config.repeated_times; ite++) {
-    // Fork
-    counter.load(std::memory_order_seq_cst);
-    for (size_t i = 0; i != performance_test_config_.run_config.concurrent_session_runs; ++i) {
-      counter++;
-      tpool->Schedule([this, &counter, &m, &cv]() {
-        session_->ThreadSafeRun();
-        // Simplified version of Eigen::Barrier
-        std::lock_guard<std::mutex> lg(m);
-        counter--;
-        cv.notify_all();
-      });
-    }
+  // Fork
+  counter.load(std::memory_order_seq_cst);
+  requests.load(std::memory_order_seq_cst);
+
+  for (size_t i = 0; i != run_config.concurrent_session_runs; ++i) {
```

Comment:
Why do you removed one level of loop?

Reply:
This way we do the number of requests that the '-r' parameter specified. It's confusing to run with something like '-r 1000 -c 2' and have that send 2000 requests vs '-r 1000 -c 4' and have that send 4000 requests. Maybe 'concurrent_session_runs' should be renamed to 'concurrent_requests'.